### PR TITLE
Initial pass "hardening" against macro and ADL issues

### PR DESCRIPTION
**Pull request by rodgert**
The gist of this change is to rename all public template parameters, arguments, and locals in line with libstdc++ "uglification" protocols to avoid issues with preprocessor macros colliding with our internal identifier names.

Also, the using namespace declarations have been adjusted to avoid ADL name collision issues.

** Please note this is the first set of changes of this nature
This *must* be applied to *every* implementation file. I have held off applying these changes to the rest of include/pstl/internal because I want to avoid any major merge conflicts with backend updates.

I would like to amend this pull request with the rest of the changes, but I need to request that you hold off on *any* concurrent changes to files under include/pstl until this pull req is complete and accepted. 

Also Note, I'm not sure why git insisted on registering "empty" merges for the pull req you have already accepted, but this is baselined against that merge from upstream.

### CODE REVIEW

File: include/pstl/internal/glue_algorithm_defs.h [hunk 1]
```diff
@@ -30,494 +30,494 @@ namespace std {
...
 
 // [alg.merge]
-template< class ExecutionPolicy, class ForwardIterator1, class ForwardIterator2, class ForwardIterator, class Compare>
-pstl::internal::enable_if_execution_policy<ExecutionPolicy, ForwardIterator>
-merge(ExecutionPolicy&& exec, ForwardIterator1 first1, ForwardIterator1 last1, ForwardIterator2 first2, ForwardIterator2 last2,
-      ForwardIterator d_first, Compare comp);
+template< class _ExecutionPolicy, class _ForwardIterator1, class _ForwardIterator2, class _ForwardIterator, class _Compare>
+pstl::internal::enable_if_execution_policy<_ExecutionPolicy, _ForwardIterator>
+merge(_ExecutionPolicy&& __exec, _ForwardIterator1 __first1, _ForwardIterator1 __last1, _ForwardIterator2 __first2, _ForwardIterator2 __last2,
+      _ForwardIterator d_first, _Compare __comp);
```
Comment (capatober) [inline include/pstl/internal/glue_algorithm_defs.h]: Should 2 underscores be added before d_first here and below?

File: include/pstl/internal/glue_memory_defs.h [hunk 2]
```diff
@@ -27,63 +27,63 @@ namespace std {
...
+template<class _ExecutionPolicy, class _ForwardIterator, class _Tp>
+pstl::internal::enable_if_execution_policy<_ExecutionPolicy, void>
+uninitialized_fill(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last, const _Tp& __value);
 
-template<class ExecutionPolicy, class ForwardIterator, class Size, class T>
-pstl::internal::enable_if_execution_policy<ExecutionPolicy, ForwardIterator>
-uninitialized_fill_n(ExecutionPolicy&& exec, ForwardIterator first, Size n, const T& value);
+template<class _ExecutionPolicy, class _ForwardIterator, class _Size, class _Tp>
+pstl::internal::enable_if_execution_policy<_ExecutionPolicy, _ForwardIterator>
+uninitialized_fill_n(_ExecutionPolicy&& __exec, _ForwardIterator __first, _Size n, const _Tp& __value);
```
Comment (capatober) [inline include/pstl/internal/glue_memory_defs.h]: 2 underscores before n

File: include/pstl/internal/glue_numeric_defs.h [hunk 3]
```diff
@@ -25,83 +25,83 @@
...
+                         _UnaryOperation __unary_op);
 
 // [adjacent.difference]
 
-template<class ExecutionPolicy, class ForwardIterator1, class ForwardIterator2, class BinaryOperation>
-pstl::internal::enable_if_execution_policy<ExecutionPolicy, ForwardIterator2>
-adjacent_difference(ExecutionPolicy&& exec, ForwardIterator1 first, ForwardIterator1 last, ForwardIterator2 d_first, BinaryOperation op);
+template<class _ExecutionPolicy, class _ForwardIterator1, class _ForwardIterator2, class _BinaryOperation>
+pstl::internal::enable_if_execution_policy<_ExecutionPolicy, _ForwardIterator2>
+adjacent_difference(_ExecutionPolicy&& __exec, _ForwardIterator1 __first, _ForwardIterator1 __last, _ForwardIterator2 d_first, _BinaryOperation op);
```
Comment (capatober) [inline include/pstl/internal/glue_numeric_defs.h]: 2 underscores before d_first

**Comment by rodgert** [comment]
I am about to issue an update to this pull request. Which includes the rest of the PSTL files. I will do another pass through to verify consistency. I fixed some of these issues while working on the next set of changes

**Comment by rodgert** [comment]
I've now pushed what I believe is the remaining uglification and ADL hardening required before this can be integrated into libstdc++

File: include/pstl/internal/unseq_backend_simd.h [hunk 4]
```diff
@@ -29,464 +29,470 @@
...
 }
 
-template<class InputIterator, class DifferenceType, class BinaryPredicate>
-DifferenceType simd_calc_mask_2(InputIterator first, DifferenceType n, bool* __restrict mask, BinaryPredicate pred) noexcept {
-    DifferenceType count = 0;
+template<class _InputIterator, class _DifferenceType, class BinaryPredicate>
+_DifferenceType simd_calc_mask_2(_InputIterator __first, _DifferenceType __n, bool* __restrict __mask, BinaryPredicate __pred) noexcept {
+    _DifferenceType __count = 0;
 
 __PSTL_PRAGMA_SIMD_REDUCTION(+:count)
```
Comment (capatober) [inline include/pstl/internal/unseq_backend_simd.h]: There is an error "‘count’ has not been declared".
I suppose "__count" should be here.

File: include/pstl/internal/unseq_backend_simd.h [hunk 5]
```diff
@@ -29,464 +29,470 @@
...
-    while ( last - first >= block_size ) {
-        difference_type found = 0;
+    const _difference_type __block_size = 8;
+    alignas(64) _difference_type __lane[__block_size] = {0};
+    while ( __last - __first >= __block_size ) {
+        _difference_type __found = 0;
 __PSTL_PRAGMA_VECTOR_UNALIGNED // Do not generate peel loop part
-__PSTL_PRAGMA_SIMD_REDUCTION(|:found)
-        for ( i = 0; i < block_size-1; ++i ) {
+__PSTL_PRAGMA_SIMD_REDUCTION(|:_found)
```
Comment (capatober) [inline include/pstl/internal/unseq_backend_simd.h]: There is an error "‘_found’ has not been declared".
I suppose "__found" should be here.

File: include/pstl/internal/unseq_backend_simd.h [hunk 6]
```diff
@@ -29,464 +29,470 @@
...
-    difference_type i = 0;
+    typedef typename std::iterator_traits<_Index>::difference_type _difference_type;
+    _difference_type __i = 0;
 
 #if __PSTL_EARLYEXIT_PRESENT
     //Some compiler versions fail to compile the following loop when iterators are used. Indices are used instead
-    const difference_type n = last-first-1;
-__PSTL_PRAGMA_VECTOR_UNALIGNED
+    const _difference_type __n = __last - __first-1;
+__PSTL_PRAGMA_VECT__OR_UNALIGNED
```
Comment (capatober) [inline include/pstl/internal/unseq_backend_simd.h]: __PSTL_PRAGMA_VECTOR_UNALIGNED should be here

**Comment by rodgert** [comment]
Should be fixed now

File: include/pstl/internal/algorithm_impl.h [hunk 7]
```diff
@@ -133,186 +138,197 @@ void pattern_walk_brick( Iterator first, Iterator last, Brick brick, /*parallel=
...
+template<class _RandomAccessIterator, class _Function>
+void brick_it_walk1( _RandomAccessIterator __first, _RandomAccessIterator __last, _Function __f, /*vector=*/std::true_type ) noexcept {
+    unseq_backend::simd_it_walk_1(__first, __last - __first, __f);
 }
 
-template<class Iterator, class Function, class IsVector>
-void pattern_it_walk1( Iterator first, Iterator last, Function f, IsVector is_vector, /*parallel=*/std::false_type ) noexcept {
-    brick_it_walk1( first, last, f, is_vector );
+template<class __ForwardIterator, class _Function, class _IsVector>
+void pattern_it_walk1( __ForwardIterator __first, __ForwardIterator __last, _Function __f, _IsVector __is_vector,
```
Comment (capatober) [inline include/pstl/internal/algorithm_impl.h]: __ForwardIterator --> _ForwardIterator (and below as well)

File: include/pstl/internal/algorithm_impl.h [hunk 8]
```diff
@@ -133,186 +138,197 @@ void pattern_walk_brick( Iterator first, Iterator last, Brick brick, /*parallel=
...
 }
 
 //------------------------------------------------------------------------
 // walk1_n
 //------------------------------------------------------------------------
-template<class InputIterator, class Size, class Function>
-InputIterator brick_walk1_n(InputIterator first, Size n, Function f, /*IsVectorTag=*/std::false_type ) {
-    return internal::for_each_n( first, n, f ); // calling serial version
+template<class InputIterator, class _Size, class _Function>
+InputIterator brick_walk1_n(InputIterator __first, _Size __n, _Function __f, /*_IsVectorTag=*/std::false_type ) {
```
Comment (capatober) [inline include/pstl/internal/algorithm_impl.h]: Missing underscores before InputIterator and RandomAccessIterator until brick_walk2

File: include/pstl/internal/algorithm_impl.h [hunk 9]
```diff
@@ -133,186 +138,197 @@ void pattern_walk_brick( Iterator first, Iterator last, Brick brick, /*parallel=
...
+        return __first2 + (__last1-__first1);
     });
 }
 
-template<class Iterator1, class Size, class Iterator2, class Function, class IsVector>
-Iterator2 pattern_walk2_n( Iterator1 first1, Size n, Iterator2 first2, Function f, IsVector is_vector, /*parallel=*/std::false_type ) noexcept {
-    return brick_walk2_n(first1, n, first2, f, is_vector);
+template<class _ForwardIterator1, class _Size, class _ForwardIterator2, class Function, class IsVector>
+_ForwardIterator2 pattern_walk2_n( _ForwardIterator1 __first1, _Size n, _ForwardIterator2 __first2, Function f,
+                                   IsVector is_vector, /*parallel=*/std::false_type ) noexcept {
```
Comment (capatober) [inline include/pstl/internal/algorithm_impl.h]: _Function, _IsVector (here and below)

File: include/pstl/internal/algorithm_impl.h [hunk 10]
```diff
@@ -321,149 +337,151 @@ Iterator2 pattern_walk2_brick_n( Iterator1 first1, Size n, Iterator2 first2, Bri
...
 //
 // walk3 evaluates f(x,y,z) for (x,y,z) drawn from [first1,last1), [first2,...), [first3,...)
 //------------------------------------------------------------------------
-template<class Iterator1, class Iterator2, class Iterator3, class Function>
-Iterator3 brick_walk3( Iterator1 first1, Iterator1 last1, Iterator2 first2, Iterator3 first3, Function f, /*vector=*/std::false_type ) noexcept {
-    for(; first1!=last1; ++first1, ++first2, ++first3 )
-        f(*first1, *first2, *first3);
-    return first3;
+template<class _ForwardIterator1, class _ForwardIterator2, class _ForwardIterator3, class _Function>
+_ForwardIterator3 brick_walk3( _ForwardIterator1 __first1, _ForwardIterator1 last1, _ForwardIterator2 __first2, _ForwardIterator3 __first3, _Function __f, /*vector=*/std::false_type ) noexcept {
```
Comment (capatober) [inline include/pstl/internal/algorithm_impl.h]: last1 --> __last1

File: include/pstl/internal/algorithm_impl.h [hunk 11]
```diff
@@ -474,805 +492,800 @@ ForwardIterator pattern_find_if(ForwardIterator first, ForwardIterator last, Pre
...
-    return std::copy_if(first, last, result, pred);
+template<class _ForwardIterator, class OutputIterator, class _UnaryPredicate>
+OutputIterator brick_copy_if(_ForwardIterator __first, _ForwardIterator __last, OutputIterator __result, _UnaryPredicate __pred, /*vector=*/std::false_type) noexcept {
+    return std::copy_if(__first, __last, __result, __pred);
 }
 
-template<class InputIterator, class OutputIterator, class UnaryPredicate>
-OutputIterator brick_copy_if(InputIterator first, InputIterator last, OutputIterator result, UnaryPredicate pred, /*vector=*/std::true_type) noexcept {
+template<class _ForwardIterator, class OutputIterator, class _UnaryPredicate>
+OutputIterator brick_copy_if(_ForwardIterator __first, _ForwardIterator __last, OutputIterator __result, _UnaryPredicate __pred, /*vector=*/std::true_type) noexcept {
```
Comment (capatober) [inline include/pstl/internal/algorithm_impl.h]: OutputIterator --> _OutputIterator (here and below in this file)

File: include/pstl/internal/algorithm_impl.h [hunk 12]
```diff
@@ -1410,387 +1424,387 @@ ForwardIt pattern_adjacent_find(ForwardIt first, ForwardIt last, BinaryPredicate
...
 }
 
-template<class ForwardIterator, class T, class IsVector>
-ForwardIterator pattern_fill(ForwardIterator first, ForwardIterator last, const T& value, /*is_parallel=*/std::true_type, IsVector vec) {
-    return except_handler([first, last, &value, vec]() {
-        par_backend::parallel_for(first, last, [&value, vec](ForwardIterator begin, ForwardIterator end) {
-            brick_fill(begin, end, value, vec); });
-        return last;
+template<class _ForwardIterator, class T, class _IsVector>
+_ForwardIterator pattern_fill(_ForwardIterator __first, _ForwardIterator __last, const T& __value, /*is_parallel=*/std::true_type, _IsVector __is_vector) {
```
Comment (capatober) [inline include/pstl/internal/algorithm_impl.h]: T --> _Tp

File: include/pstl/internal/algorithm_impl.h [hunk 13]
```diff
@@ -1410,387 +1424,387 @@ ForwardIt pattern_adjacent_find(ForwardIt first, ForwardIt last, BinaryPredicate
...
 
 //------------------------------------------------------------------------
 // includes
 //------------------------------------------------------------------------
 
-template<class InputIterator1, class InputIterator2, class Compare>
-bool brick_includes(InputIterator1 first1, InputIterator1 last1, InputIterator2 first2, InputIterator2 last2, Compare comp, /* IsVector = */ std::false_type) noexcept {
-    return std::includes(first1, last1, first2, last2, comp);
+template<class ForwardIterator1, class ForwardIterator2, class _Compare>
+bool brick_includes(ForwardIterator1 __first1, ForwardIterator1 __last1, ForwardIterator2 __first2, ForwardIterator2 __last2, _Compare __comp, /* _IsVector = */ std::false_type) noexcept {
```
Comment (capatober) [inline include/pstl/internal/algorithm_impl.h]: I'd suggest to add an underscore before ForwardIterator1, ForwardIterator2

File: include/pstl/internal/unseq_backend_simd.h [hunk 14]
```diff
@@ -29,464 +29,470 @@
...
 
-template<class Index, class DifferenceType, class Pred>
-DifferenceType simd_count(Index first, DifferenceType n, Pred pred) noexcept {
-    DifferenceType count = 0;
-__PSTL_PRAGMA_SIMD_REDUCTION(+:count)
-    for (DifferenceType i = 0; i < n; ++i)
-        if (pred(*(first + i)))
-            ++count;
+template<class _Index, class _DifferenceType, class _Pred>
+_DifferenceType simd_count(_Index __index, _DifferenceType __n, _Pred pred) noexcept {
```
Comment (capatober) [inline include/pstl/internal/unseq_backend_simd.h]: pred --> __pred

File: include/pstl/internal/unseq_backend_simd.h [hunk 15]
```diff
@@ -29,464 +29,470 @@
...
+    return __count;
 }
 
-template<class InputIterator, class DifferenceType, class OutputIterator, class BinaryPredicate>
-OutputIterator simd_unique_copy(InputIterator first, DifferenceType n, OutputIterator result, BinaryPredicate pred) noexcept {
-    if (n == 0)
-        return result;
+template<class _InputIterator, class _DifferenceType, class _OutputIterator, class __BinaryPredicate>
+_OutputIterator simd_unique_copy(_InputIterator __first, _DifferenceType __n, _OutputIterator __result,
+                                 __BinaryPredicate __pred) noexcept {
```
Comment (capatober) [inline include/pstl/internal/unseq_backend_simd.h]: __BinaryPredicate --> _BinaryPredicate

File: include/pstl/internal/unseq_backend_simd.h [hunk 16]
```diff
@@ -29,464 +29,470 @@
...
+    for (_DifferenceType __i = 0; __i < __n; ++__i)
+        __assigner(__first + __i, __result + __i);
+    return __result + __n;
 }
 
-template<class InputIterator, class DifferenceType, class OutputIterator, class UnaryPredicate>
-OutputIterator simd_copy_if(InputIterator first, DifferenceType n, OutputIterator result, UnaryPredicate pred) noexcept {
-    DifferenceType cnt = 0;
+template<class _InputIterator, class _DifferenceType, class _OutputIterator, class UnaryPredicate>
+_OutputIterator simd_copy_if(_InputIterator __first, _DifferenceType __n, _OutputIterator __result, UnaryPredicate __pred) noexcept {
```
Comment (capatober) [inline include/pstl/internal/unseq_backend_simd.h]: Missing underscore before UnaryPredicate (and below as well)

File: include/pstl/internal/unseq_backend_simd.h [hunk 17]
```diff
@@ -29,464 +29,470 @@
...
-        return last; // according to the standard
+    const _difference_type __n1 = __last - __first;
+    const _difference_type __n2 = __s_last - __s_first;
+    if (__n1 == 0 || __n2 == 0) {
+        return __last; // according to the standard
     }
 
     // Common case
-    // If first sequence larger than second then we'll run simd_first with parameters of first sequence.
+    // If __first sequence larger than second then we'll run simd___first with parameters of __first sequence.
```
Comment (capatober) [inline include/pstl/internal/unseq_backend_simd.h]: simd___first --> simd_first

File: include/pstl/internal/algorithm_impl.h [hunk 18]
```diff
@@ -474,805 +492,800 @@ ForwardIterator pattern_find_if(ForwardIterator first, ForwardIterator last, Pre
...
-                    [&m](difference_type total) {m=total;});
-                return result + m;
+                    [&__m](_difference_type __total) {__m = __total;});
+                return __result + __m;
             });
         }
     }
-    // Out of memory or trivial sequence - use serial algorithm
-    return brick_copy_if(first, last, result, pred, is_vector);
+    // _Out of memory or trivial sequence - use serial algorithm
```
Comment (capatober) [inline include/pstl/internal/algorithm_impl.h]: _Out --> Out

File: include/pstl/internal/glue_algorithm_defs.h [hunk 19]
```diff
@@ -30,494 +30,494 @@ namespace std {
...
+unique_copy(_executionpolicy&& __exec, _forwarditerator1 __first, _forwarditerator1 __last, _forwarditerator2 __result);
 
 // [alg.reverse]
 
-template<class ExecutionPolicy, class BidirectionalIterator>
-pstl::internal::enable_if_execution_policy<ExecutionPolicy, void>
-reverse(ExecutionPolicy&& exec, BidirectionalIterator first, BidirectionalIterator last);
+template<class _executionpolicy, class _bidirectionaliterator>
+__pstl::internal::enable_if_execution_policy<_executionpolicy, void>
+reverse(_executionpolicy&& __exec, _bidirectionaliterator __first, _bidirectionaliterator __last);
```
[the remark this answers is not in the document]
Comment (rodgert) [reply include/pstl/internal/glue_algorithm_defs.h]: It looks like I managed to somehow fat-fingure lowercasing these en-masse

File: include/pstl/internal/parallel_backend_utils.h [hunk 20]
```diff
@@ -24,104 +24,109 @@
...
     }
 }
 
-template<typename Buf>
+template<typename _Buf>
 class stack {
-    typedef typename std::iterator_traits<decltype(Buf(0).get())>::value_type T;
-    typedef typename std::iterator_traits<T*>::difference_type difference_type;
+    typedef typename std::iterator_traits<decltype(_Buf(0).get())>::value_type _value_type;
+    typedef typename std::iterator_traits<_value_type*>::difference_type _difference_type;
```
[the remark this answers is not in the document]
Comment (rodgert) [reply include/pstl/internal/parallel_backend_utils.h]: Yes, I re-reviewed the practice ln libstdc++ and _ValueType, _DistanceType are more consistent for function local typedefs.

**Comment by rodgert** [comment]
The most recent commit should address the mistted/misnamed identifiers. I will create a separate commit that captures renaming the local typedefs and submit that later today.

File: include/pstl/internal/glue_algorithm_defs.h [hunk 21]
```diff
@@ -133,117 +133,117 @@ search_n(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator _
...
+template<class _ExecutionPolicy, class _ForwardIterator1, class _ForwardIterator2>
+__pstl::internal::enable_if_execution_policy<_ExecutionPolicy,_ForwardIterator2>
+copy(_ExecutionPolicy&& __exec, _ForwardIterator1 __first, _ForwardIterator1 __last, _ForwardIterator2 __result);
 
-template<class _executionpolicy, class _forwarditerator1, class _size, class _forwarditerator2>
-__pstl::internal::enable_if_execution_policy<_executionpolicy,_forwarditerator2>
-copy_n(_executionpolicy&& __exec, _forwarditerator1 __first, _size __n, _forwarditerator2 __result);
+template<class _ExecutionPolicy, class _ForwardIterator1, class _size, class _ForwardIterator2>
+__pstl::internal::enable_if_execution_policy<_ExecutionPolicy,_ForwardIterator2>
+copy_n(_ExecutionPolicy&& __exec, _ForwardIterator1 __first, _size __n, _ForwardIterator2 __result);
```
Comment (capatober) [inline include/pstl/internal/glue_algorithm_defs.h]: Should _size be replaced with _Size here?

File: include/pstl/internal/glue_algorithm_defs.h [hunk 22]
```diff
@@ -133,117 +133,117 @@ search_n(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator _
...
           _binaryoperation __op);
 
 // [alg.replace]
 
-template<class _executionpolicy, class _forwarditerator, class _unarypredicate, class _tp>
-__pstl::internal::enable_if_execution_policy<_executionpolicy, void>
-replace_if(_executionpolicy&& __exec, _forwarditerator __first, _forwarditerator __last, _unarypredicate __pred, const _tp& __new_value);
+template<class _ExecutionPolicy, class _ForwardIterator, class _UnaryPredicate, class _tp>
+__pstl::internal::enable_if_execution_policy<_ExecutionPolicy, void>
+replace_if(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last, _UnaryPredicate __pred, const _tp& __new_value);
```
Comment (capatober) [inline include/pstl/internal/glue_algorithm_defs.h]: Should _tp replaced with _Tp here and below?

**Pull request merged**